File: binarytrees/binarytrees.go
// -*- mode: go; coding: utf-8; indent-tabs-mode: nil; c-basic-offset: 4 -*-
// vim: set ft=c fenc=utf-8 ts=4 sw=4 et :

// go build -o binarytrees-go binarytrees.go

package main

import "fmt"
import "strconv"
import "os"

type Node struct {
    item int
    left, right *Node
}

func (node *Node) Check() int {
    if node.left == nil {
        return node.item
    }
    return node.item + node.left.Check() - node.right.Check()
}

func create(item int, depth int) *Node {
    if depth == 0 {
        return &Node{0, nil, nil}
    }
    return &Node{item, create(2 * item - 1, depth - 1), create(2 * item, depth - 1)}
}

func main() {
    n := 0
    if len(os.Args) > 1 { n, _ = strconv.Atoi(os.Args[1]) }
    minDepth := 4
    maxDepth := minDepth + 2
    if maxDepth < n {
        maxDepth = n
    }
    stretchDepth := maxDepth + 1

    check := create(0, stretchDepth).Check()
    fmt.Printf("stretch tree of depth %d\t check: %d\n", stretchDepth, check)

    longLivedTree := create(0, maxDepth)

    for depth := minDepth; depth <= maxDepth; depth += 2 {
        iterations := 1 << uint(maxDepth - depth + minDepth)
        check = 0

        for i := 1; i <= iterations; i++ {
            check += create(i, depth).Check()
            check += create(-i, depth).Check()
        }
        fmt.Printf("%d\t trees of depth %d\t check: %d\n", iterations, depth, check)
    }

    fmt.Printf("long lived tree of depth %d\t check: %d\n", maxDepth, longLivedTree.Check())
}

File: binarytrees/binarytrees.nim
# -*- mode: nim; coding: utf-8; indent-tabs-mode: nil; c-basic-offset: 2 -*-
# vim: set ft=c fenc=utf-8 ts=2 sw=2 et :

# nim c -d:release -out:binarytrees-nim binarytrees.nim
# time ./binarytrees-nim 20

import "os"
import "strutils"

type
  Node = ref object
    item: int
    left, right: Node

proc check(self: Node): int =
  if self.left == nil:
    return self.item
  return self.item + self.left.check() - self.right.check()

proc create(item, depth: int): Node =
  if depth == 0:
    return Node(item: item, left: nil, right: nil)
  return Node(item: item, left: create(2 * item - 1, depth - 1), right: create(2 * item, depth - 1))

# main

let n = if paramCount() > 0: paramStr(1).parseInt else: 0
let minDepth = 4
let maxDepth = if minDepth + 2 < n: n else: minDepth + 2
let stretchDepth = maxDepth + 1

echo "stretch tree of depth ", stretchDepth, "\t check: ", create(0, stretchDepth).check()

let longLivedTree = create(0, maxDepth)

for depth in countup(minDepth, maxDepth, 2):
  let iterations = 1 shl (maxDepth - depth + minDepth)
  var check = 0
  for i in 1..iterations:
    check += create(i, depth).check()
    check += create(-i, depth).check()
  echo 2 * iterations, "\t trees of depth ", depth, "\t check: ", check

echo "long lived tree of depth ", maxDepth, "\t check: ", longLivedTree.check()
